File: ios/App/App/ZipCentralDirectoryInspector.swift
import Foundation

struct PackArchiveManifest: Decodable {
    struct ArchiveIdentity: Decodable {
        let bytes: Int
        let name: String
        let sha256: String
    }

    struct Ceilings: Decodable {
        let compressedBytes: Int
        let extractedBytes: Int
        let fileCount: Int
    }

    struct FileRecord: Decodable {
        let bytes: Int
        let path: String
        let sha256: String
    }

    let allowedExtensions: [String]
    let archive: ArchiveIdentity
    let ceilings: Ceilings
    let files: [FileRecord]
    let packId: String
    let requiredEntitlementId: String
    let schemaVersion: Int
    let version: String
}

struct InspectedPackInventory {
    struct Entry {
        let path: String
        let compressedBytes: Int
        let extractedBytes: Int
        let sha256: String
    }

    let entries: [Entry]
    let compressedBytes: Int
    let extractedBytes: Int
}

enum PackInspectionError: Error {
    case rejected
}

enum PackCapabilityValidator {
    private static let safeIdentifier = try! NSRegularExpression(
        pattern: "^[a-z0-9][a-z0-9._-]{0,63}$"
    )
    private static let archiveName = try! NSRegularExpression(
        pattern: "^[a-z0-9][a-z0-9._-]{0,119}\\.zip$"
    )

    static func validateIdentifier(_ value: String) throws {
        let range = NSRange(value.startIndex..<value.endIndex, in: value)
        guard safeIdentifier.firstMatch(in: value, range: range)?.range == range else {
            throw PackInspectionError.rejected
        }
    }

    static func validateArchiveName(_ value: String) throws {
        let range = NSRange(value.startIndex..<value.endIndex, in: value)
        guard archiveName.firstMatch(in: value, range: range)?.range == range else {
            throw PackInspectionError.rejected
        }
    }

    static func validateCapabilityURL(
        _ capability: String,
        packId: String,
        version: String,
        archiveName: String
    ) throws -> URL {
        try validateIdentifier(packId)
        try validateIdentifier(version)
        try validateArchiveName(archiveName)
        guard capability.utf8.count <= 8_192,
              let components = URLComponents(string: capability),
              components.scheme == "https",
              components.host == "b3-gateway.eugnel.uk",
              components.user == nil,
              components.password == nil,
              components.port == nil,
              components.fragment == nil else {
            throw PackInspectionError.rejected
        }
        let expectedPath = "/v1/packs/\(packId)/\(version)/\(archiveName)"
        guard components.percentEncodedPath == expectedPath,
              let query = components.percentEncodedQuery else {
            throw PackInspectionError.rejected
        }
        let queryExpression = try! NSRegularExpression(
            pattern: "^expires=([1-9][0-9]*)&cap=([A-Za-z0-9_-]{43})$"
        )
        let range = NSRange(query.startIndex..<query.endIndex, in: query)
        guard let match = queryExpression.firstMatch(in: query, range: range),
              match.range == range,
              let expiresRange = Range(match.range(at: 1), in: query),
              let expires = UInt64(query[expiresRange]),
              expires > 0,
              let url = components.url,
              url.absoluteString == capability,
              url.absoluteString == "https://b3-gateway.eugnel.uk\(expectedPath)?\(query)" else {
            throw PackInspectionError.rejected
        }
        return url
    }
}

struct ValidatedPackRange {
    let status: Int
    let startByte: Int
    let endByteExclusive: Int
    let totalBytes: Int
    let etag: String
}

enum PackRangeResponseValidator {
    static func validate(
        statusCode: Int,
        responseURL: URL?,
        expectedURL: URL,
        etag: String?,
        contentRange: String?,
        bodyBytes: Int,
        requestedStart: Int,
        requestedEndExclusive: Int
    ) throws -> ValidatedPackRange {
        guard responseURL?.absoluteString == expectedURL.absoluteString,
              let etag,
              !etag.isEmpty,
              etag.utf8.count <= 256,
              bodyBytes > 0,
              bodyBytes <= 1_048_576 else {
            throw PackInspectionError.rejected
        }
        if statusCode == 200 {
            return ValidatedPackRange(
                status: 200,
                startByte: 0,
                endByteExclusive: bodyBytes,
                totalBytes: bodyBytes,
                etag: etag
            )
        }
        guard statusCode == 206, let contentRange else {
            throw PackInspectionError.rejected
        }
        let expression = try! NSRegularExpression(pattern: "^bytes ([0-9]+)-([0-9]+)/([1-9][0-9]*)$")
        let range = NSRange(contentRange.startIndex..<contentRange.endIndex, in: contentRange)
        guard let match = expression.firstMatch(in: contentRange, range: range),
              match.range == range,
              let startRange = Range(match.range(at: 1), in: contentRange),
              let endRange = Range(match.range(at: 2), in: contentRange),
              let totalRange = Range(match.range(at: 3), in: contentRange),
              let start = Int(contentRange[startRange]),
              let inclusiveEnd = Int(contentRange[endRange]),
              let total = Int(contentRange[totalRange]),
              start == requestedStart,
              inclusiveEnd >= start,
              inclusiveEnd < total,
              inclusiveEnd + 1 == min(requestedEndExclusive, total),
              bodyBytes == inclusiveEnd - start + 1 else {
            throw PackInspectionError.rejected
        }
        return ValidatedPackRange(
            status: 206,
            startByte: start,
            endByteExclusive: inclusiveEnd + 1,
            totalBytes: total,
            etag: etag
        )
    }
}

enum ZipCentralDirectoryInspector {
    private static let endSignature: UInt32 = 0x06054b50
    private static let centralSignature: UInt32 = 0x02014b50
    private static let localSignature: UInt32 = 0x04034b50
    private static let utf8Flag: UInt16 = 0x0800
    private static let regularMode: UInt16 = 0o100644
    private static let allowedExtensions = Set([".json", ".m4a"])

    private struct RangeRecord {
        let start: Int
        let end: Int
    }

    static func inspect(
        archiveURL: URL,
        manifest: PackArchiveManifest
    ) throws -> InspectedPackInventory {
        guard manifest.schemaVersion == 1,
              manifest.ceilings.fileCount > 0,
              manifest.ceilings.fileCount <= 16,
              manifest.ceilings.compressedBytes > 0,
              manifest.ceilings.compressedBytes <= 1_048_576,
              manifest.ceilings.extractedBytes > 0,
              manifest.ceilings.extractedBytes <= 4_194_304,
              Set(manifest.allowedExtensions) == allowedExtensions else {
            throw PackInspectionError.rejected
        }
        let data = try Data(contentsOf: archiveURL, options: [.mappedIfSafe])
        guard data.count == manifest.archive.bytes,
              data.count <= manifest.ceilings.compressedBytes,
              data.count >= 22 else {
            throw PackInspectionError.rejected
        }
        let bytes = [UInt8](data)
        let endOffsets = signatureOffsets(endSignature, in: bytes)
        guard endOffsets.count == 1, let endOffset = endOffsets.first,
              endOffset + 22 == bytes.count,
              read16(bytes, endOffset + 4) == 0,
              read16(bytes, endOffset + 6) == 0,
              read16(bytes, endOffset + 8) == read16(bytes, endOffset + 10),
              read16(bytes, endOffset + 20) == 0 else {
            throw PackInspectionError.rejected
        }
        let entryCount = Int(read16(bytes, endOffset + 10))
        let centralSize = Int(read32(bytes, endOffset + 12))
        let centralOffset = Int(read32(bytes, endOffset + 16))
        guard entryCount > 0,
              entryCount <= manifest.ceilings.fileCount,
              centralOffset >= 0,
              centralSize >= 0,
              centralOffset.addingReportingOverflow(centralSize).overflow == false,
              centralOffset + centralSize == endOffset else {
            throw PackInspectionError.rejected
        }

        let declared = try declaredFiles(manifest)
        var seenPaths = Set<String>()
        var foldedPaths = Set<String>()
        var localOffsets = Set<Int>()
        var localRanges: [RangeRecord] = []
        var dataRanges: [RangeRecord] = []
        var entries: [InspectedPackInventory.Entry] = []
        var compressedTotal = 0
        var extractedTotal = 0
        var cursor = centralOffset

        for _ in 0..<entryCount {
            guard cursor + 46 <= endOffset,
                  read32(bytes, cursor) == centralSignature else {
                throw PackInspectionError.rejected
            }
            let madeBy = read16(bytes, cursor + 4)
            let versionNeeded = read16(bytes, cursor + 6)
            let centralFlags = read16(bytes, cursor + 8)
            let centralMethod = read16(bytes, cursor + 10)
            let centralCRC = read32(bytes, cursor + 16)
            let compressedBytes = Int(read32(bytes, cursor + 20))
            let extractedBytes = Int(read32(bytes, cursor + 24))
            let nameLength = Int(read16(bytes, cursor + 28))
            let extraLength = Int(read16(bytes, cursor + 30))
            let commentLength = Int(read16(bytes, cursor + 32))
            let diskStart = read16(bytes, cursor + 34)
            let internalAttributes = read16(bytes, cursor + 36)
            let externalAttributes = read32(bytes, cursor + 38)
            let localOffset = Int(read32(bytes, cursor + 42))
            let recordEnd = cursor + 46 + nameLength + extraLength + commentLength
            guard madeBy >> 8 == 3,
                  versionNeeded <= 20,
                  centralFlags == utf8Flag,
                  centralMethod == 0 || centralMethod == 8,
                  compressedBytes != Int(UInt32.max),
                  extractedBytes != Int(UInt32.max),
                  nameLength > 0,
                  extraLength == 0,
                  commentLength == 0,
                  diskStart == 0,
                  internalAttributes == 0,
                  UInt16(externalAttributes >> 16) == regularMode,
                  recordEnd <= endOffset,
                  localOffset < centralOffset,
                  localOffsets.insert(localOffset).inserted else {
                throw PackInspectionError.rejected
            }
            let centralNameBytes = Array(bytes[(cursor + 46)..<(cursor + 46 + nameLength)])
            guard let path = String(bytes: centralNameBytes, encoding: .utf8) else {
                throw PackInspectionError.rejected
            }
            try validatePath(path, manifest: manifest)
            let folded = path.precomposedStringWithCanonicalMapping.lowercased()
            guard seenPaths.insert(path).inserted,
                  foldedPaths.insert(folded).inserted,
                  let declaration = declared[path],
                  declaration.bytes == extractedBytes else {
                throw PackInspectionError.rejected
            }

            guard localOffset + 30 <= centralOffset,
                  read32(bytes, localOffset) == localSignature else {
                throw PackInspectionError.rejected
            }
            let localNameLength = Int(read16(bytes, localOffset + 26))
            let localExtraLength = Int(read16(bytes, localOffset + 28))
            let dataStart = localOffset + 30 + localNameLength + localExtraLength
            let dataEnd = dataStart.addingReportingOverflow(compressedBytes)
            guard read16(bytes, localOffset + 4) == versionNeeded,
                  read16(bytes, localOffset + 6) == centralFlags,
                  read16(bytes, localOffset + 8) == centralMethod,
                  read16(bytes, localOffset + 10) == read16(bytes, cursor + 12),
                  read16(bytes, localOffset + 12) == read16(bytes, cursor + 14),
                  read32(bytes, localOffset + 14) == centralCRC,
                  read32(bytes, localOffset + 18) == UInt32(compressedBytes),
                  read32(bytes, localOffset + 22) == UInt32(extractedBytes),
                  localNameLength == nameLength,
                  localExtraLength == 0,
                  dataEnd.overflow == false,
                  dataEnd.partialValue <= centralOffset,
                  Array(bytes[(localOffset + 30)..<(localOffset + 30 + localNameLength)]) == centralNameBytes else {
                throw PackInspectionError.rejected
            }
            localRanges.append(RangeRecord(start: localOffset, end: dataEnd.partialValue))
            dataRanges.append(RangeRecord(start: dataStart, end: dataEnd.partialValue))
            compressedTotal = try checkedTotal(compressedTotal, compressedBytes)
            extractedTotal = try checkedTotal(extractedTotal, extractedBytes)
            entries.append(.init(
                path: path,
                compressedBytes: compressedBytes,
                extractedBytes: extractedBytes,
                sha256: declaration.sha256
            ))
            cursor = recordEnd
        }

        guard cursor == endOffset,
              entries.count == declared.count,
              compressedTotal <= manifest.ceilings.compressedBytes,
              extractedTotal <= manifest.ceilings.extractedBytes else {
            throw PackInspectionError.rejected
        }
        try requireTiledLocalRecords(localRanges, centralOffset: centralOffset)
        try requireNoOverlap(dataRanges)
        return InspectedPackInventory(
            entries: entries,
            compressedBytes: compressedTotal,
            extractedBytes: extractedTotal
        )
    }

    private static func declaredFiles(
        _ manifest: PackArchiveManifest
    ) throws -> [String: PackArchiveManifest.FileRecord] {
        guard manifest.files.count <= manifest.ceilings.fileCount else {
            throw PackInspectionError.rejected
        }
        var result: [String: PackArchiveManifest.FileRecord] = [:]
        var folded = Set<String>()
        for file in manifest.files {
            try validatePath(file.path, manifest: manifest)
            guard file.bytes >= 0,
                  file.sha256.range(of: "^[0-9a-f]{64}$", options: .regularExpression) != nil,
                  result[file.path] == nil,
                  folded.insert(file.path.precomposedStringWithCanonicalMapping.lowercased()).inserted else {
                throw PackInspectionError.rejected
            }
            result[file.path] = file
        }
        return result
    }

    private static func validatePath(
        _ path: String,
        manifest: PackArchiveManifest
    ) throws {
        guard !path.isEmpty,
              path == path.precomposedStringWithCanonicalMapping,
              path.utf8.allSatisfy({ $0 < 128 }),
              path.range(
                of: "^[A-Za-z0-9][A-Za-z0-9._-]*(/[A-Za-z0-9][A-Za-z0-9._-]*)*$",
                options: .regularExpression
              ) != nil,
              !path.hasPrefix("/"),
              !path.contains("\\"),
              !path.hasSuffix("/"),
              !path.split(separator: "/", omittingEmptySubsequences: false).contains(where: {
                  $0.isEmpty || $0 == "." || $0 == ".." || $0.hasPrefix(".")
              }),
              let dot = path.lastIndex(of: ".") else {
            throw PackInspectionError.rejected
        }
        let suffix = String(path[dot...])
        guard allowedExtensions.contains(suffix), manifest.allowedExtensions.contains(suffix) else {
            throw PackInspectionError.rejected
        }
    }

    private static func checkedTotal(_ total: Int, _ value: Int) throws -> Int {
        guard value >= 0 else { throw PackInspectionError.rejected }
        let sum = total.addingReportingOverflow(value)
        guard !sum.overflow else { throw PackInspectionError.rejected }
        return sum.partialValue
    }

    private static func requireTiledLocalRecords(
        _ ranges: [RangeRecord],
        centralOffset: Int
    ) throws {
        let sorted = ranges.sorted { $0.start < $1.start }
        var cursor = 0
        for range in sorted {
            guard range.start == cursor, range.end > range.start else {
                throw PackInspectionError.rejected
            }
            cursor = range.end
        }
        guard cursor == centralOffset else { throw PackInspectionError.rejected }
    }

    private static func requireNoOverlap(_ ranges: [RangeRecord]) throws {
        let sorted = ranges.sorted { $0.start < $1.start }
        for index in 1..<sorted.count where sorted[index].start < sorted[index - 1].end {
            throw PackInspectionError.rejected
        }
    }

    private static func signatureOffsets(_ signature: UInt32, in bytes: [UInt8]) -> [Int] {
        guard bytes.count >= 4 else { return [] }
        return (0...(bytes.count - 4)).filter { read32(bytes, $0) == signature }
    }

    private static func read16(_ bytes: [UInt8], _ offset: Int) -> UInt16 {
        UInt16(bytes[offset]) | (UInt16(bytes[offset + 1]) << 8)
    }

    private static func read32(_ bytes: [UInt8], _ offset: Int) -> UInt32 {
        UInt32(bytes[offset]) |
            (UInt32(bytes[offset + 1]) << 8) |
            (UInt32(bytes[offset + 2]) << 16) |
            (UInt32(bytes[offset + 3]) << 24)
    }
}

File: android/app/src/main/java/uk/eugnel/ks2spelling/ZipCentralDirectoryInspector.java
package uk.eugnel.ks2spelling;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.zip.CRC32;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

/** Owns every ZIP metadata decision before platform extraction is allowed. */
public final class ZipCentralDirectoryInspector {
    static final int MAXIMUM_FILE_COUNT = 1_024;
    static final int MAXIMUM_COMPRESSED_BYTES = 32 * 1_024 * 1_024;
    static final int MAXIMUM_EXTRACTED_BYTES = 32 * 1_024 * 1_024;

    private static final long END_SIGNATURE = 0x06054b50L;
    private static final long CENTRAL_SIGNATURE = 0x02014b50L;
    private static final long LOCAL_SIGNATURE = 0x04034b50L;
    private static final int UTF8_FLAG = 0x0800;
    private static final int REGULAR_MODE = 0100644;
    private static final Set<String> ALLOWED_EXTENSIONS = Collections.unmodifiableSet(
        new HashSet<>(Arrays.asList(".json", ".m4a"))
    );
    private static final Pattern SAFE_PATH = Pattern.compile(
        "^[A-Za-z0-9][A-Za-z0-9._-]*(/[A-Za-z0-9][A-Za-z0-9._-]*)*$"
    );

    private ZipCentralDirectoryInspector() {}

    static void validateManifestCeilings(ManifestInventory manifest) throws IOException {
        require(manifest != null && manifest.schemaVersion == 1);
        require(manifest.archiveBytes > 0 && manifest.archiveBytes <= MAXIMUM_COMPRESSED_BYTES);
        require(manifest.fileCountCeiling > 0
            && manifest.fileCountCeiling <= MAXIMUM_FILE_COUNT);
        require(manifest.compressedBytesCeiling > 0
            && manifest.compressedBytesCeiling <= MAXIMUM_COMPRESSED_BYTES);
        require(manifest.extractedBytesCeiling > 0
            && manifest.extractedBytesCeiling <= MAXIMUM_EXTRACTED_BYTES);
        require(new HashSet<>(manifest.allowedExtensions).equals(ALLOWED_EXTENSIONS));
    }

    public static Inventory inspect(byte[] bytes, ManifestInventory manifest) throws IOException {
        validateManifestCeilings(manifest);
        require(bytes.length == manifest.archiveBytes
            && bytes.length <= manifest.compressedBytesCeiling && bytes.length >= 22);

        List<Integer> endOffsets = signatureOffsets(END_SIGNATURE, bytes);
        require(endOffsets.size() == 1);
        int endOffset = endOffsets.get(0);
        require(endOffset + 22 == bytes.length);
        require(read16(bytes, endOffset + 4) == 0 && read16(bytes, endOffset + 6) == 0);
        require(read16(bytes, endOffset + 8) == read16(bytes, endOffset + 10));
        require(read16(bytes, endOffset + 20) == 0);
        int entryCount = read16(bytes, endOffset + 10);
        long centralSizeLong = read32(bytes, endOffset + 12);
        long centralOffsetLong = read32(bytes, endOffset + 16);
        require(entryCount > 0 && entryCount <= manifest.fileCountCeiling);
        require(centralSizeLong <= Integer.MAX_VALUE && centralOffsetLong <= Integer.MAX_VALUE);
        int centralSize = (int) centralSizeLong;
        int centralOffset = (int) centralOffsetLong;
        require(checkedAdd(centralOffset, centralSize) == endOffset);

        Map<String, FileDeclaration> declared = declaredFiles(manifest);
        Set<String> seenPaths = new HashSet<>();
        Set<String> foldedPaths = new HashSet<>();
        Set<Integer> localOffsets = new HashSet<>();
        List<RangeRecord> localRanges = new ArrayList<>();
        List<RangeRecord> dataRanges = new ArrayList<>();
        List<Entry> entries = new ArrayList<>();
        int compressedTotal = 0;
        int extractedTotal = 0;
        int cursor = centralOffset;

        for (int index = 0; index < entryCount; index += 1) {
            require(checkedAdd(cursor, 46) <= endOffset && read32(bytes, cursor) == CENTRAL_SIGNATURE);
            int madeBy = read16(bytes, cursor + 4);
            int versionNeeded = read16(bytes, cursor + 6);
            int centralFlags = read16(bytes, cursor + 8);
            int method = read16(bytes, cursor + 10);
            long crc = read32(bytes, cursor + 16);
            long compressedLong = read32(bytes, cursor + 20);
            long extractedLong = read32(bytes, cursor + 24);
            int nameLength = read16(bytes, cursor + 28);
            int extraLength = read16(bytes, cursor + 30);
            int commentLength = read16(bytes, cursor + 32);
            int diskStart = read16(bytes, cursor + 34);
            int internalAttributes = read16(bytes, cursor + 36);
            long externalAttributes = read32(bytes, cursor + 38);
            long localOffsetLong = read32(bytes, cursor + 42);
            require(compressedLong != 0xffffffffL && extractedLong != 0xffffffffL);
            require(compressedLong <= Integer.MAX_VALUE && extractedLong <= Integer.MAX_VALUE);
            require(localOffsetLong <= Integer.MAX_VALUE);
            int compressed = (int) compressedLong;
            int extracted = (int) extractedLong;
            int localOffset = (int) localOffsetLong;
            int recordEnd = checkedAdd(cursor, checkedAdd(46, checkedAdd(nameLength, checkedAdd(extraLength, commentLength))));
            require((madeBy >>> 8) == 3 && versionNeeded <= 20);
            require(centralFlags == UTF8_FLAG && (method == 0 || method == 8));
            require(nameLength > 0 && extraLength == 0 && commentLength == 0);
            require(diskStart == 0 && internalAttributes == 0);
            require(((externalAttributes >>> 16) & 0xffffL) == REGULAR_MODE);
            require(recordEnd <= endOffset && localOffset < centralOffset && localOffsets.add(localOffset));
            byte[] centralNameBytes = slice(bytes, cursor + 46, checkedAdd(cursor + 46, nameLength));
            String path = strictAsciiUtf8(centralNameBytes);
            validatePath(path, manifest);
            String folded = Normalizer.normalize(path, Normalizer.Form.NFC).toLowerCase(Locale.ROOT);
            FileDeclaration declaration = declared.get(path);
            require(seenPaths.add(path) && foldedPaths.add(folded));
            require(declaration != null && declaration.bytes == extracted);

            require(checkedAdd(localOffset, 30) <= centralOffset && read32(bytes, localOffset) == LOCAL_SIGNATURE);
            int localNameLength = read16(bytes, localOffset + 26);
            int localExtraLength = read16(bytes, localOffset + 28);
            int dataStart = checkedAdd(localOffset, checkedAdd(30, checkedAdd(localNameLength, localExtraLength)));
            int dataEnd = checkedAdd(dataStart, compressed);
            require(read16(bytes, localOffset + 4) == versionNeeded);
            require(read16(bytes, localOffset + 6) == centralFlags);
            require(read16(bytes, localOffset + 8) == method);
            require(read16(bytes, localOffset + 10) == read16(bytes, cursor + 12));
            require(read16(bytes, localOffset + 12) == read16(bytes, cursor + 14));
            require(read32(bytes, localOffset + 14) == crc);
            require(read32(bytes, localOffset + 18) == compressedLong);
            require(read32(bytes, localOffset + 22) == extractedLong);
            require(localNameLength == nameLength && localExtraLength == 0 && dataEnd <= centralOffset);
            require(java.util.Arrays.equals(
                slice(bytes, localOffset + 30, checkedAdd(localOffset + 30, localNameLength)),
                centralNameBytes
            ));
            localRanges.add(new RangeRecord(localOffset, dataEnd));
            dataRanges.add(new RangeRecord(dataStart, dataEnd));
            compressedTotal = checkedAdd(compressedTotal, compressed);
            extractedTotal = checkedAdd(extractedTotal, extracted);
            entries.add(new Entry(path, method, crc, compressed, extracted, declaration.sha256));
            cursor = recordEnd;
        }

        require(cursor == endOffset && entries.size() == declared.size());
        require(compressedTotal <= manifest.compressedBytesCeiling);
        require(extractedTotal <= manifest.extractedBytesCeiling);
        requireTiledLocalRecords(localRanges, centralOffset);
        requireNoOverlap(dataRanges);
        return new Inventory(entries, compressedTotal, extractedTotal);
    }

    /** Consumes only the byte array already approved by {@link #inspect(byte[], ManifestInventory)}. */
    public static int consumeVerifiedEntries(byte[] archiveBytes, Inventory inventory,
        int extractedBytesCeiling, VerifiedEntryConsumer consumer) throws Exception {
        require(archiveBytes != null && inventory != null && consumer != null);
        require(extractedBytesCeiling > 0
            && extractedBytesCeiling <= MAXIMUM_EXTRACTED_BYTES);
        int extractedTotal = 0;
        try (ZipInputStream input = new ZipInputStream(new ByteArrayInputStream(archiveBytes))) {
            for (Entry approved : inventory.entries) {
                ZipEntry platform = input.getNextEntry();
                require(platform != null && !platform.isDirectory());
                require(approved.path.equals(platform.getName()));
                require(platform.getMethod() == approved.method);
                require(platform.getCrc() == approved.crc32);
                require(platform.getSize() == approved.extractedBytes);
                require(platform.getCompressedSize() == approved.compressedBytes);

                MessageDigest digest = MessageDigest.getInstance("SHA-256");
                CRC32 crc = new CRC32();
                ByteArrayOutputStream content = new ByteArrayOutputStream(
                    Math.min(approved.extractedBytes, 64 * 1024)
                );
                byte[] buffer = new byte[64 * 1024];
                int written = 0;
                while (true) {
                    int count = input.read(buffer);
                    if (count == -1) break;
                    require(count > 0);
                    written = checkedAdd(written, count);
                    require(written <= approved.extractedBytes);
                    digest.update(buffer, 0, count);
                    crc.update(buffer, 0, count);
                    content.write(buffer, 0, count);
                }
                input.closeEntry();
                require(written == approved.extractedBytes);
                require(crc.getValue() == approved.crc32);
                require(hex(digest.digest()).equals(approved.sha256));
                require(platform.getCrc() == approved.crc32);
                require(platform.getSize() == approved.extractedBytes);
                require(platform.getCompressedSize() == approved.compressedBytes);
                extractedTotal = checkedAdd(extractedTotal, written);
                require(extractedTotal <= extractedBytesCeiling);
                consumer.accept(approved, content.toByteArray());
            }
            require(input.getNextEntry() == null);
            require(input.read() == -1);
        }
        require(extractedTotal == inventory.extractedBytes);
        return extractedTotal;
    }

    private static Map<String, FileDeclaration> declaredFiles(ManifestInventory manifest) throws IOException {
        require(!manifest.files.isEmpty() && manifest.files.size() <= manifest.fileCountCeiling);
        Map<String, FileDeclaration> result = new HashMap<>();
        Set<String> folded = new HashSet<>();
        for (FileDeclaration file : manifest.files) {
            validatePath(file.path, manifest);
            require(file.bytes >= 0 && file.sha256.matches("^[0-9a-f]{64}$"));
            require(result.put(file.path, file) == null);
            require(folded.add(Normalizer.normalize(file.path, Normalizer.Form.NFC).toLowerCase(Locale.ROOT)));
        }
        return result;
    }

    private static void validatePath(String path, ManifestInventory manifest) throws IOException {
        require(path != null && !path.isEmpty() && path.equals(Normalizer.normalize(path, Normalizer.Form.NFC)));
        require(StandardCharsets.US_ASCII.newEncoder().canEncode(path) && SAFE_PATH.matcher(path).matches());
        require(!path.startsWith("/") && !path.contains("\\") && !path.endsWith("/"));
        String[] segments = path.split("/", -1);
        for (String segment : segments) require(!segment.isEmpty() && !segment.equals(".") && !segment.equals("..") && !segment.startsWith("."));
        int dot = path.lastIndexOf('.');
        require(dot >= 0);
        String extension = path.substring(dot);
        require(ALLOWED_EXTENSIONS.contains(extension) && manifest.allowedExtensions.contains(extension));
    }

    private static String strictAsciiUtf8(byte[] bytes) throws IOException {
        for (byte value : bytes) require((value & 0x80) == 0);
        String value = new String(bytes, StandardCharsets.UTF_8);
        require(java.util.Arrays.equals(value.getBytes(StandardCharsets.UTF_8), bytes));
        return value;
    }

    private static void requireTiledLocalRecords(List<RangeRecord> ranges, int centralOffset) throws IOException {
        ranges.sort(Comparator.comparingInt(value -> value.start));
        int cursor = 0;
        for (RangeRecord range : ranges) {
            require(range.start == cursor && range.end > range.start);
            cursor = range.end;
        }
        require(cursor == centralOffset);
    }

    private static void requireNoOverlap(List<RangeRecord> ranges) throws IOException {
        ranges.sort(Comparator.comparingInt(value -> value.start));
        for (int index = 1; index < ranges.size(); index += 1) {
            require(ranges.get(index).start >= ranges.get(index - 1).end);
        }
    }

    private static List<Integer> signatureOffsets(long signature, byte[] bytes) {
        List<Integer> result = new ArrayList<>();
        for (int offset = 0; offset <= bytes.length - 4; offset += 1) {
            if (read32(bytes, offset) == signature) result.add(offset);
        }
        return result;
    }

    private static int checkedAdd(int left, int right) throws IOException {
        try { return Math.addExact(left, right); } catch (ArithmeticException error) { throw rejected(); }
    }

    private static byte[] slice(byte[] bytes, int start, int end) throws IOException {
        require(start >= 0 && end >= start && end <= bytes.length);
        return java.util.Arrays.copyOfRange(bytes, start, end);
    }

    private static String hex(byte[] bytes) {
        StringBuilder output = new StringBuilder(bytes.length * 2);
        for (byte value : bytes) output.append(String.format(Locale.ROOT, "%02x", value & 0xff));
        return output.toString();
    }

    private static int read16(byte[] bytes, int offset) {
        if (offset < 0 || offset + 2 > bytes.length) return -1;
        return (bytes[offset] & 0xff) | ((bytes[offset + 1] & 0xff) << 8);
    }

    private static long read32(byte[] bytes, int offset) {
        if (offset < 0 || offset + 4 > bytes.length) return -1;
        return ((long) bytes[offset] & 0xffL)
            | (((long) bytes[offset + 1] & 0xffL) << 8)
            | (((long) bytes[offset + 2] & 0xffL) << 16)
            | (((long) bytes[offset + 3] & 0xffL) << 24);
    }

    private static void require(boolean condition) throws IOException {
        if (!condition) throw rejected();
    }

    private static IOException rejected() { return new IOException("Pack archive rejected."); }

    private static final class RangeRecord {
        final int start;
        final int end;
        RangeRecord(int start, int end) { this.start = start; this.end = end; }
    }

    public static final class FileDeclaration {
        public final String path;
        public final int bytes;
        public final String sha256;
        public FileDeclaration(String path, int bytes, String sha256) {
            this.path = path;
            this.bytes = bytes;
            this.sha256 = sha256;
        }
    }

    public static final class ManifestInventory {
        public final int schemaVersion;
        public final int archiveBytes;
        public final int fileCountCeiling;
        public final int compressedBytesCeiling;
        public final int extractedBytesCeiling;
        public final List<String> allowedExtensions;
        public final List<FileDeclaration> files;
        public ManifestInventory(int schemaVersion, int archiveBytes, int fileCountCeiling,
            int compressedBytesCeiling, int extractedBytesCeiling,
            List<String> allowedExtensions, List<FileDeclaration> files) {
            this.schemaVersion = schemaVersion;
            this.archiveBytes = archiveBytes;
            this.fileCountCeiling = fileCountCeiling;
            this.compressedBytesCeiling = compressedBytesCeiling;
            this.extractedBytesCeiling = extractedBytesCeiling;
            this.allowedExtensions = Collections.unmodifiableList(new ArrayList<>(allowedExtensions));
            this.files = Collections.unmodifiableList(new ArrayList<>(files));
        }
    }

    public static final class Entry {
        public final String path;
        public final int method;
        public final long crc32;
        public final int compressedBytes;
        public final int extractedBytes;
        public final String sha256;
        Entry(String path, int method, long crc32, int compressedBytes, int extractedBytes,
            String sha256) {
            this.path = path;
            this.method = method;
            this.crc32 = crc32;
            this.compressedBytes = compressedBytes;
            this.extractedBytes = extractedBytes;
            this.sha256 = sha256;
        }
    }

    @FunctionalInterface
    public interface VerifiedEntryConsumer {
        void accept(Entry entry, byte[] content) throws Exception;
    }

    public static final class Inventory {
        public final List<Entry> entries;
        public final int compressedBytes;
        public final int extractedBytes;
        Inventory(List<Entry> entries, int compressedBytes, int extractedBytes) {
            this.entries = Collections.unmodifiableList(new ArrayList<>(entries));
            this.compressedBytes = compressedBytes;
            this.extractedBytes = extractedBytes;
        }
    }
}
